lib/core/sdam/monitor.js

```diff
@@ -0,0 +1,247 @@
+          bson: server.s.bson,
+          connectionType: Connection
+        },
+        server.s.options,
+
+        // force BSON serialization options
+        {
+          raw: false,
+          promoteLongs: true,
+          promoteValues: true,
+          promoteBuffers: true
+        }
+      )
+    );
+  }
+
+  connect() {
+    if (this.s.state !== STATE_CLOSED) {
+      return;
+    }
+
+    monitorServer(this);
+  }
+
+  requestCheck() {
+    if (INVALID_REQUEST_CHECK_STATES.has(this.s.state)) {
+      return;
+    }
+
+    const heartbeatFrequencyMS = this.options.heartbeatFrequencyMS;
+    const minHeartbeatFrequencyMS = this.options.minHeartbeatFrequencyMS;
+    const remainingTime = heartbeatFrequencyMS - calculateDurationInMs(this[kLastCheckTime]);
+    if (remainingTime > minHeartbeatFrequencyMS && this[kMonitorId]) {
+      clearTimeout(this[kMonitorId]);
+      rescheduleMonitoring(this, minHeartbeatFrequencyMS);
+      return;
+    }
+
+    if (this[kMonitorId]) {
+      clearTimeout(this[kMonitorId]);
+    }
+
+    monitorServer(this);
+  }
+
+  close() {
+    if (this.s.state === STATE_CLOSED || this.s.state === STATE_CLOSING) {
+      return;
+    }
+
+    stateTransition(this, STATE_CLOSING);
+    this[kCancellationToken].emit('cancel');
+    if (this[kMonitorId]) {
+      clearTimeout(this[kMonitorId]);
+    }
+
+    if (this[kConnection]) {
+      this[kConnection].destroy({ force: true });
+    }
+
+    this.emit('close');
+    stateTransition(this, STATE_CLOSED);
+  }
+}
+
+function checkServer(monitor, callback) {
+  if (monitor[kConnection] && monitor[kConnection].closed) {
+    monitor[kConnection] = undefined;
+  }
+
+  monitor.emit('serverHeartbeatStarted', new ServerHeartbeatStartedEvent(monitor.address));
+
+  if (monitor[kConnection] != null) {
+    const connectTimeoutMS = monitor.options.connectTimeoutMS;
+    monitor[kConnection].command(
+      'admin.$cmd',
+      { ismaster: true },
+      { socketTimeout: connectTimeoutMS },
+      (err, result) => {
+        if (err) {
+          return callback(err);
+        }
+
+        return callback(undefined, result.result);
+      }
+    );
+
+    return;
+  }
+
+  // connecting does an implicit `ismaster`
+  connect(monitor.connectOptions, monitor[kCancellationToken], (err, conn) => {
+    if (err) {
+      monitor[kConnection] = undefined;
+      callback(err);
+      return;
+    }
+
+    if (monitor.s.state === STATE_CLOSING || monitor.s.state === STATE_CLOSED) {
+      conn.destroy({ force: true });
+      callback(new MongoError('monitor was destroyed'));
+      return;
+    }
+
+    monitor[kConnection] = conn;
+    callback(undefined, conn.description);
+  });
+}
+
+function monitorServer(monitor) {
+  const start = process.hrtime();
+  stateTransition(monitor, STATE_MONITORING);
+
+  // TODO: the next line is a legacy event, remove in v4
+  process.nextTick(() => monitor.emit('monitoring', monitor[kServer]));
```

Comment:
🤔 is there a more spec way to listen for a `monitoring` event? Use case that comes to mind is logging in compass.

Reply:
yep, you can use [SDAM event monitoring](https://github.com/mongodb/node-mongodb-native/blob/master/lib/core/sdam/topology.js#L90-L92) (or [this link](https://mongodb.github.io/node-mongodb-native/3.4/reference/management/sdam-monitoring/)) to get these details. This custom `monitoring` event will be removed in v4.